libs/utils/include/utils/JobSystem.h

```diff
@@ -354,19 +354,56 @@ class JobSystem {
         assert(job);
         size_t index = job - mJobStorageBase;
         assert(index >= 0 && index < MAX_JOB_COUNT);
+
+        // put the job into the queue first
         workQueue.push(uint16_t(index + 1));
+        // then increase our active job count
+        uint32_t oldActiveJobs = mActiveJobs.fetch_add(1, std::memory_order_relaxed);
+        // but it's possible that the job has already been picked-up, so oldActiveJobs could be
+        // negative for instance. We signal only if that's not the case.
+        if (oldActiveJobs >= 0) {
+            wakeOne(); // wake-up a thread if needed...
+        }
     }
 
     Job* pop(WorkQueue& workQueue) noexcept {
+        // decrement mActiveJobs first, this is to ensure that if there is only a single job left
+        // (and we're about to pick it up), other threads don't loop trying to do the same.
+        mActiveJobs.fetch_sub(1, std::memory_order_relaxed);
```

Comment:
Can you end up with mActiveJobs < 0 here but workQueue.pop() returning a valid job?

Reply:
Yes, you could but it wouldn't be a problem. This would mean than another thread is for e.g. doing a steal() at the same time. So, if we do get a valid job from the queue, this means that the other thread didn't, and it will "fix" the counter by doing a fetch_add.
On the other hand, if we don't get a job, then we are fixing the counter.

The key is to do the correction based on whether we actually got a job or not, NOT based on the value of the counter.

The other key is to signal if needed when fixing the counter.